vendor/bat-native-ads/src/bat/ads/internal/ml/pipeline/text_processing/embedding_processing.cc

```diff
@@ -0,0 +1,114 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "bat/ads/internal/ml/pipeline/text_processing/embedding_processing.h"
+
+#include <algorithm>
+#include <cstdint>
+#include <utility>
+#include <vector>
+
+#include "base/base64.h"
+#include "base/check.h"
+#include "base/strings/string_split.h"
+#include "base/strings/string_util.h"
+#include "base/values.h"
+#include "bat/ads/internal/base/crypto/crypto_util.h"
+#include "bat/ads/internal/base/logging_util.h"
+#include "bat/ads/internal/ml/data/vector_data.h"
+#include "bat/ads/internal/ml/pipeline/pipeline_embedding_info.h"
+#include "bat/ads/internal/ml/pipeline/pipeline_embedding_value_util.h"
+#include "bat/ads/internal/ml/pipeline/text_processing/embedding_info.h"
+
+namespace ads::ml::pipeline {
+
+// static
+std::unique_ptr<EmbeddingProcessing> EmbeddingProcessing::CreateFromValue(
+    base::Value resource_value,
+    std::string* error_message) {
+  DCHECK(error_message);
+
+  auto embedding_processing = std::make_unique<EmbeddingProcessing>();
+  if (!embedding_processing->SetEmbeddingPipeline(std::move(resource_value))) {
+    *error_message = "Failed to parse embedding pipeline JSON";
+    return nullptr;
+  }
+
+  return embedding_processing;
+}
+
+EmbeddingProcessing::EmbeddingProcessing() = default;
+
+EmbeddingProcessing::~EmbeddingProcessing() = default;
+
+bool EmbeddingProcessing::IsInitialized() const {
+  return is_initialized_;
+}
+
+bool EmbeddingProcessing::SetEmbeddingPipeline(base::Value resource_value) {
+  base::Value::Dict* value = resource_value.GetIfDict();
+  if (!value) {
+    return is_initialized_;
+  }
+
+  absl::optional<EmbeddingPipelineInfo> embedding_pipeline =
+      EmbeddingPipelineFromValue(*value);
+  if (!embedding_pipeline) {
+    is_initialized_ = false;
+  } else {
+    embedding_pipeline_ = *embedding_pipeline;
+    is_initialized_ = true;
+  }
+
+  return is_initialized_;
+}
+
+TextEmbeddingInfo EmbeddingProcessing::EmbedText(
+    const std::string& text) const {
+  if (!IsInitialized()) {
+    return {};
+  }
+
+  if (text.empty()) {
+    return {};
+  }
+
+  const std::vector<float> embedding_zeroed(embedding_pipeline_.dim, 0.0F);
+  const VectorData embedding_vector_data = VectorData(embedding_zeroed);
+  TextEmbeddingInfo text_embedding;
+  text_embedding.embedding = embedding_vector_data;
+
+  const std::vector<std::string> tokens = base::SplitString(
+      text, " ", base::TRIM_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
+  std::vector<std::string> in_vocab_tokens;
+
+  for (const auto& token : tokens) {
```

Comment:
Do we have a rough idea of max tokens?

Reply:
@tmancey This is dependent on the text from the webpage we are embedding, so it's difficult to give a known max.

Reply:
We must run some tests on heavy, medium sites as this could cause performance issues. Also, we need to consider a max limit for the token count, which ideally is configurable via Griffin. We do the same with browser history.

Reply:
Discussed and as this is only og:title this is fine for now, however, if we use other tag names in the future we should reconsider perf